File: finance-tracker-backend/src/main/java/com/finance/tracker/accounts/domain/AccountCreateUpdateRequest.java
package com.finance.tracker.accounts.domain;

import com.fasterxml.jackson.annotation.JsonAlias;
import jakarta.validation.constraints.NotNull;

import java.math.BigDecimal;

public record AccountCreateUpdateRequest(
        @JsonAlias({"name", "accountName"})
        String accountName,
        @JsonAlias({"type", "accountType"})
        @NotNull
        AccountType accountType,
        String lastFour,
        String institution,
        String currency,
        @JsonAlias({"balance", "startingBalance"})
        BigDecimal startingBalance,       // asset only
        @JsonAlias({"balance", "currentOutstanding"})
        BigDecimal currentOutstanding,    // liability only
        int statementDayOfMonth,         // credit card only
        int dueDayOfMonth,            // credit card only
        BigDecimal creditLimit,           // credit card only
        String notes,
        @NotNull(message = "category should be present. value is either ASSET or LIABILITY")
        AccountCategory category           // ASSET / LIABILITY
) {
    public AccountCreateUpdateRequest(
            String accountName,
            AccountType accountType,
            String lastFour,
            String currency,
            BigDecimal startingBalance,
            BigDecimal currentOutstanding,
            int statementDayOfMonth,
            int dueDayOfMonth,
            BigDecimal creditLimit,
            String notes,
            AccountCategory category
    ) {
        this(accountName, accountType, lastFour, null, currency, startingBalance, currentOutstanding, statementDayOfMonth, dueDayOfMonth, creditLimit, notes, category);
    }

    public String name() {
        return accountName != null ? accountName : "";
    }

    public AccountType type() {
        return accountType;
    }

    public BigDecimal balance() {
        return startingBalance != null ? startingBalance : (currentOutstanding != null ? currentOutstanding : BigDecimal.ZERO);
    }
}

File: finance_app/lib/features/account/data/model/account_create_update_request.dart
class AccountCreateUpdateRequest {
  final String accountName;
  final int lastFour;
  final String accountType;
  final DateTime openingDate;
  final double startingBalance;
  final double currentOutstanding;
  final String currency;
  final double creditLimit;
  final String statementDayOfMonth;
  final String dueDayOfMonth;
  final String notes;
  final bool hideFromSelection;
  final bool hideFromReports;
  final String category; // 'asset' or 'liability'

  AccountCreateUpdateRequest({
    required this.accountName,
    required this.lastFour,
    required this.accountType,
    required this.openingDate,
    required this.startingBalance,
    required this.currentOutstanding,
    required this.currency,
    required this.creditLimit,
    required this.statementDayOfMonth,
    required this.dueDayOfMonth,
    required this.notes,
    required this.hideFromSelection,
    required this.hideFromReports,
    required this.category,
  });

  Map<String, dynamic> toJson() {
    return {
      'accountName': accountName,
      'lastFour': lastFour,
      'accountType': accountType,
      'openingDate': openingDate.toIso8601String(),
      'startingBalance': startingBalance,
      'currentOutstanding': currentOutstanding,
      'currency': currency,
      'creditLimit': creditLimit,
      'statementDayOfMonth': statementDayOfMonth,
      'dueDayOfMonth': dueDayOfMonth,
      'notes': notes,
      'hideFromSelection': hideFromSelection,
      'hideFromReports': hideFromReports,
      'category': category,
    };
  }
}
